File: src/host/grammar.py
"""Map the name-first launcher grammar onto the internal command-first form.

Both launchers accept `LAUNCHER NAME COMMAND [SUBCOMMAND] [PARAMETERS]`. Only
build, list and update --all omit NAME. Command names are reserved so the first
argument is never ambiguous, and old command-first forms fail with a hint.
"""
import re
import shlex
import sys

HELP = ('help', '--help', '-h')
GLOBAL = ('build', 'list', 'update')
SESSIONS = ('copilot', 'claude', 'codex', 'hermes', 'opencode', 'deepseek', 't3')
SANDBOX = ('up', 'start', 'stop', 'restart', 'remove', 'shell', 'ssh-config', 'check', 'check-full',
           'fingerprint', 'agents', 'tools', 'run', 'tool', 'service', 'forward', 'update', *SESSIONS)
RESERVED = frozenset((*HELP, *GLOBAL, *SANDBOX, 'azdo'))


def command_line(prog, args):
    return ' '.join([prog, *(arg if arg in ('NAME', 'COMMAND') else shlex.quote(arg) for arg in args)])


def removed_azdo(prog, name='NAME'):
    return ValueError('The azdo --pat-env command was removed. Save a PAT for the sandbox with: '
                      + command_line(prog, [name, 'tools', 'setup', 'azdo', '--persist']))


def update_names(args):
    """Positional sandbox names in old-style update arguments."""
    names, skip = [], False
    for arg in args:
        if skip:
            skip = False
        elif arg == '--capabilities':
            skip = True
        elif not arg.startswith('-'):
            names.append(arg)
    return names


def normalize(args, prog):
    """Return [COMMAND, NAME, *PARAMETERS] for sandbox commands, or [COMMAND, *PARAMETERS]."""
    if not args or args[0] in HELP:
        return ['help']
    first, rest = args[0], list(args[1:])
    if first in RESERVED and rest and rest[0] in SANDBOX:
        raise ValueError(f"'{first}' is a command name and cannot be used as a sandbox name.")
    if first in ('build', 'list'):
        return [first, *rest]
    if first == 'update':
        if '--all' in rest or any(arg in HELP for arg in rest):
            return [first, *rest]
        names = update_names(rest)
        options = [arg for arg in rest if arg not in names]
        hints = [command_line(prog, [name, 'update', *options]) for name in names]
        raise ValueError('Update one sandbox with its name first, or all of them with --all: '
                         + ', '.join(hints or [command_line(prog, ['NAME', 'update'])])
                         + f', {command_line(prog, ["update", "--all"])}')
    if first == 'azdo':
        raise removed_azdo(prog, rest[0] if rest and not rest[0].startswith('-') else 'NAME')
    if first in SANDBOX:
        name, parameters = (rest[0], rest[1:]) if rest and not rest[0].startswith('-') else ('NAME', rest)
        raise ValueError('Commands take the sandbox name first: '
                         + command_line(prog, [name, first, *parameters]))
    if not re.fullmatch(r'[a-zA-Z0-9][a-zA-Z0-9_.-]*', first):
        raise ValueError('Use an alphanumeric container name (plus _, ., -).')
    if not rest:
        raise ValueError(f'Use {command_line(prog, ["NAME", "COMMAND"])} [SUBCOMMAND] [PARAMETERS]. '
                         f'Run {prog} --help.')
    command, parameters = rest[0], rest[1:]
    if command == 'azdo':
        raise removed_azdo(prog, first)
    if command in ('build', 'list'):
        raise ValueError(f'{command_line(prog, [command])} does not take a sandbox name.')
    if command == 'update' and '--all' in parameters:
        raise ValueError(f'Use {command_line(prog, ["update", "--all"])} without a sandbox name.')
    if command == 'update' and update_names(parameters):
        raise ValueError(f'Update one sandbox per command: {command_line(prog, [first, "update"])} [--no-build] '
                         '[--capabilities LIST], or update --all.')
    if command not in SANDBOX:
        raise ValueError(f'Unknown command: {command}. Run {prog} --help.')
    return [command, first, *parameters]


if __name__ == '__main__':
    # Print NUL-terminated arguments so Bash can read them back without quoting.
    try:
        sys.stdout.write(''.join(f'{arg}\0' for arg in normalize(sys.argv[2:], sys.argv[1])))
    except ValueError as error:
        sys.exit(f'Error: {error}')

File: internal/cli/grammar.go
package cli

import (
	"fmt"
	"regexp"
	"slices"
	"strings"
)

var namedCommands = strings.Fields("up start stop restart remove shell ssh-config check check-full fingerprint agents tools run tool service forward update adopt copilot claude codex hermes opencode deepseek t3")
var unnamedCommands = []string{"version", "build", "list", "update", "adopt"}
var namePattern = regexp.MustCompile(`^[a-zA-Z0-9][a-zA-Z0-9_.-]*$`)

func isHelp(s string) bool { return s == "help" || s == "--help" || s == "-h" }

func parse(args []string) (command, name string, parameters []string, err error) {
	if len(args) == 0 || isHelp(args[0]) {
		return "help", "", nil, nil
	}
	first, rest := args[0], args[1:]
	reserved := slices.Contains(namedCommands, first) || slices.Contains(unnamedCommands, first) || first == "azdo"
	if reserved && len(rest) > 0 && slices.Contains(namedCommands, rest[0]) {
		return "", "", nil, fmt.Errorf("%q is a command name and cannot be used as a sandbox name", first)
	}
	if slices.Contains(unnamedCommands, first) {
		if (first == "update" || first == "adopt") && !slices.Contains(rest, "--all") && !slices.Contains(rest, "--help") && !slices.Contains(rest, "-h") {
			return "", "", nil, fmt.Errorf("use sandboxed-agents NAME %s, or sandboxed-agents %s --all", first, first)
		}
		if (first == "list" || first == "version") && len(rest) > 0 {
			return "", "", nil, fmt.Errorf("Usage: sandboxed-agents %s", first)
		}
		return first, "", rest, nil
	}
	if first == "azdo" {
		return "", "", nil, fmt.Errorf("the azdo --pat-env command was removed; use sandboxed-agents NAME tools setup azdo --persist")
	}
	if slices.Contains(namedCommands, first) {
		return "", "", nil, fmt.Errorf("commands take the sandbox name first: sandboxed-agents NAME %s", first)
	}
	if !namePattern.MatchString(first) {
		return "", "", nil, fmt.Errorf("use an alphanumeric container name (plus _, ., -)")
	}
	if len(rest) == 0 {
		return "", "", nil, fmt.Errorf("use sandboxed-agents NAME COMMAND; run sandboxed-agents --help")
	}
	command = rest[0]
	if isHelp(command) {
		return "help", "", nil, nil
	}
	if slices.Contains([]string{"build", "list", "version"}, command) {
		return "", "", nil, fmt.Errorf("sandboxed-agents %s does not take a sandbox name", command)
	}
	if !slices.Contains(namedCommands, command) {
		return "", "", nil, fmt.Errorf("Unknown command: %s. Run sandboxed-agents --help", command)
	}
	if (command == "update" || command == "adopt") && slices.Contains(rest[1:], "--all") {
		return "", "", nil, fmt.Errorf("use sandboxed-agents %s --all without a sandbox name", command)
	}
	return command, first, rest[1:], nil
}
